### PR TITLE
Add rel_from_table_function to R relational API

### PR DESCRIPTION
**Pull request by hannes**
cc @krlmlr @Tmonster

### CODE REVIEW

File: tools/rpkg/R/relational.R [hunk 1]
```diff
@@ -379,3 +379,17 @@ rel_from_table <- function(con, table_name, schema_name="MAIN") {
...
+#' Convert a duckdb relation from a table-producing function
+#' @param name the table function name
+#' @param positional_parameters the table function positional parameters list
+#' @param named_parameters the table function named parameters list
+#' @return a duckdb relation
+#' @noRd
+#' @examples
+#' con <- DBI::dbConnect(duckdb())
+#' DBI::dbWriteTable(con, "mtcars", mtcars)
+#' rel <- rel_from_table(con, "mtcars")
```
Comment (krlmlr) [inline tools/rpkg/R/relational.R:391]: Need to tweak example:

```suggestion
#' rel <- rel_from_table_function(con, ...)
```

File: tools/rpkg/src/relational.cpp [hunk 2]
```diff
@@ -464,10 +464,44 @@ static SEXP result_to_df(duckdb::unique_ptr<QueryResult> res) {
...
+
+[[cpp11::register]] SEXP rapi_rel_from_table_function(duckdb::conn_eptr_t con, const std::string function_name,
+                                                      list positional_parameters_sexps, list named_parameters_sexps) {
+	if (!con || !con.get() || !con->conn) {
+		stop("rel_from_table_function: Invalid connection");
+	}
+	vector<Value> positional_parameters;
+
+	for (sexp parameter_sexp : positional_parameters_sexps) {
+		if (LENGTH(parameter_sexp) != 1) {
```
Comment (krlmlr) [inline tools/rpkg/src/relational.cpp:483]: Do we have functions that expect record parameters?

File: tools/rpkg/src/relational.cpp [hunk 3]
```diff
@@ -464,10 +464,44 @@ static SEXP result_to_df(duckdb::unique_ptr<QueryResult> res) {
...
+			stop("rel_from_table_function: Need scalar parameter");
+		}
+		named_parameters[names[(const R_xlen_t)named_parameter_idx]] = RApiTypes::SexpToValue(parameter_sexp, 0);
+		named_parameter_idx++;
+	}
+
+	auto rel = con->conn->TableFunction(function_name, std::move(positional_parameters), std::move(named_parameters));
+	cpp11::writable::list prot = {};
+	return make_external_prot<RelationWrapper>("duckdb_relation", prot, std::move(rel));
```
Comment (krlmlr) [inline tools/rpkg/src/relational.cpp:506]: Maybe:

```suggestion
	return make_external<RelationWrapper>("duckdb_relation", std::move(rel));
```

File: tools/rpkg/src/relational.cpp [hunk 4]
```diff
@@ -464,10 +464,44 @@ static SEXP result_to_df(duckdb::unique_ptr<QueryResult> res) {
...
+		stop("rel_from_table_function: Named parameters need names");
+	}
+	idx_t named_parameter_idx = 0;
+	for (sexp parameter_sexp : named_parameters_sexps) {
+		if (LENGTH(parameter_sexp) != 1) {
+			stop("rel_from_table_function: Need scalar parameter");
+		}
+		named_parameters[names[(const R_xlen_t)named_parameter_idx]] = RApiTypes::SexpToValue(parameter_sexp, 0);
+		named_parameter_idx++;
+	}
```
Comment (krlmlr) [inline tools/rpkg/src/relational.cpp:502]: Use correct type? Or is there perhaps an idiomatic way to iterate over parallel vectors ("bind" or such)?

```suggestion
	R_xlen_t named_parameter_idx = 0;
	for (sexp parameter_sexp : named_parameters_sexps) {
		if (LENGTH(parameter_sexp) != 1) {
			stop("rel_from_table_function: Need scalar parameter");
		}
		named_parameters[names[named_parameter_idx]] = RApiTypes::SexpToValue(parameter_sexp, 0);
		named_parameter_idx++;
	}
```

**Review by krlmlr — COMMENTED** [review]
Nice! Should we add a test to read from a bunch of CSV files too?

File: tools/rpkg/src/relational.cpp (hunk 2 above)
Comment (hannes) [reply tools/rpkg/src/relational.cpp:483]: There are some who expect lists or structs as parameters, indeed. We should have a test that uses one of those, and the check is probably too strict.

File: tools/rpkg/src/relational.cpp (hunk 4 above)
Comment (hannes) [reply tools/rpkg/src/relational.cpp:502]: Better indeed

**Comment by hannes** [comment]
`SexpToValue` does not support lists yet, but once it does it should now work.

**Comment by krlmlr** [comment]
Are the UBSAN failures valid?

**Comment by hannes** [comment]
Yes, RE2 does not pass UBSAN. Will have to remove the test case.

**Pull request merged**